src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php

```diff
@@ -20,12 +20,54 @@
  */
 final class AttributeAutoconfigurationPass extends AbstractRecursivePass
 {
+    private $classAttributeConfigurators = [];
+    private $methodAttributeConfigurators = [];
+    private $propertyAttributeConfigurators = [];
+    private $parameterAttributeConfigurators = [];
+
     public function process(ContainerBuilder $container): void
     {
         if (80000 > \PHP_VERSION_ID || !$container->getAutoconfiguredAttributes()) {
             return;
         }
 
+        foreach ($container->getAutoconfiguredAttributes() as $attributeName => $callable) {
+            $callableReflector = new \ReflectionFunction(\Closure::fromCallable($callable));
+            if ($callableReflector->getNumberOfParameters() <= 2) {
+                $this->classAttributeConfigurators[$attributeName] = $callable;
+                continue;
+            }
+
+            $reflectorParameter = $callableReflector->getParameters()[2];
+            $parameterType = $reflectorParameter->getType();
+            $types = [];
+            if ($parameterType instanceof \ReflectionUnionType) {
+                foreach ($parameterType->getTypes() as $type) {
+                    $types[] = $type->getName();
+                }
+            } elseif ($parameterType instanceof \ReflectionNamedType) {
+                $types[] = $parameterType->getName();
+            } else {
+                throw new \LogicException('Callable passed to registerAttributeForAutoconfiguration() should have a type for parameter #3. Use one or more of the following types: \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter.');
```

Comment:
let's use LogicException from the component's namespace
let's also tell file + line info to help spot the source of the issue

Reply:
Used the correct `LogicException` now. But I'm not sure how to get the file and line info as this is executed in the `CompilerPass` and not where it's defined (the call to `registerAttributeForAutoconfiguration`)